File: frontend/src/utils/paymentRecurrence.ts
import { Payment, Student } from '../types'
import { addWeeks, addMonths, parseISO, format, endOfMonth, differenceInCalendarDays, isValid, startOfMonth } from 'date-fns'

/**
 * Gera pagamentos recorrentes a partir de um pagamento base
 * Usa janela de dados dinâmica baseada no mês atual
 * 
 * REGRA DE NEGÓCIO: Alunos com mais de 2 meses consecutivos de inadimplência
 * são considerados inativos e param de gerar novos pagamentos (realismo).
 * 
 * @param basePayment - Pagamento base com recorrência configurada
 * @returns Array de pagamentos gerados
 */
export const generateRecurringPayments = (
  basePayment: Payment,
  studentsById?: Map<string, Student>
): Payment[] => {
  // Pagamento anual antecipado: não gera recorrências
  if (basePayment.paymentFrequency === 'ANNUAL' || basePayment.paidInAdvance) {
    return [basePayment]
  }
  
  if (!basePayment.recurrence || basePayment.recurrence === 'ONCE') {
    return [basePayment]
  }

  const payments: Payment[] = []
  const startDate = parseISO(basePayment.dueDate)
  
  // Validação: se data for inválida, retorna apenas o pagamento base
  if (!isValid(startDate)) {
    console.warn(`Invalid dueDate for payment ${basePayment.id}: ${basePayment.dueDate}`)
    return [basePayment]
  }
  
  // Janela realista para demo: até o fim do mês corrente
  // (evita projeção de dezenas de pendências futuras irreais)
  let endDate = endOfMonth(new Date())

  if (startDate > endDate) {
    endDate = startDate
  }

  const student = studentsById?.get(basePayment.studentId)
  if (student && student.status !== 'ACTIVE' && student.lastPaymentDate) {
    const studentLastPaymentDate = endOfMonth(parseISO(student.lastPaymentDate))
    if (!isNaN(studentLastPaymentDate.getTime()) && studentLastPaymentDate < endDate) {
      endDate = studentLastPaymentDate
    }
  }

  if (basePayment.recurrenceEndDate) {
    const recurrenceEndDate = parseISO(basePayment.recurrenceEndDate)
    if (!isNaN(recurrenceEndDate.getTime()) && recurrenceEndDate < endDate) {
      endDate = recurrenceEndDate
    }
  }
  
  let currentDate = startDate
  let paymentIndex = 0
  let consecutiveUnpaidMonths = 0

  while (currentDate <= endDate) {
    const today = new Date()
    today.setHours(0, 0, 0, 0)
    const currentMonthStart = startOfMonth(today)
    const isPast = currentDate < today
    const delayDays = isPast ? differenceInCalendarDays(today, currentDate) : 0
    const generatedStatus: Payment['status'] = !isPast ? 'PENDING' : delayDays >= 45 ? 'DEFAULTED' : 'OVERDUE'
    const isBaseInstance = paymentIndex === 0
    
    // Determina status final do pagamento
    let finalStatus: Payment['status'] = isBaseInstance ? basePayment.status : generatedStatus

    // Realismo da demo:
    // para aluno ACTIVE, meses passados já consolidados como pagos,
    // evitando acúmulo artificial de dezenas de pendências antigas.
    if (!isBaseInstance && student?.status === 'ACTIVE') {
      finalStatus = currentDate < currentMonthStart ? 'PAID' : 'PENDING'
    }
    
    // Realismo: após 2 meses de inadimplência consecutiva, aluno é considerado inativo
    if (finalStatus === 'OVERDUE' || finalStatus === 'DEFAULTED') {
      consecutiveUnpaidMonths++
      
      // Stop generating após 2 meses sem pagar
      if (consecutiveUnpaidMonths > 2) {
        break
      }
    } else if (finalStatus === 'PAID' || finalStatus === 'PENDING') {
      // Reseta contador se pagamento foi pago ou ainda está pendente
      consecutiveUnpaidMonths = 0
    }
    
    const newPayment: Payment = {
      ...basePayment,
      id: isBaseInstance ? basePayment.id : `${basePayment.id}-${paymentIndex}`,
      dueDate: format(currentDate, 'yyyy-MM-dd'),
      paidDate: isBaseInstance ? basePayment.paidDate : undefined,
      status: finalStatus,
      createdAt: basePayment.createdAt,
    }

    payments.push(newPayment)

    // Calcula próxima data baseado na recorrência
    switch (basePayment.recurrence) {
      case 'WEEKLY':
        currentDate = addWeeks(currentDate, 1)
        break
      case 'BIWEEKLY':
        currentDate = addWeeks(currentDate, 2)
        break
      case 'MONTHLY':
        currentDate = addMonths(currentDate, 1)
        break
      default:
        // Tipo de recorrência desconhecido - força saída do loop
        console.warn(`Unknown recurrence type: ${basePayment.recurrence}`)
        return payments
    }

    paymentIndex++

    // Segurança: máximo 200 recorrências
    if (paymentIndex > 200) break
  }

  return payments
}

/**
 * Expande todos os payments com recorrência
 */
export const expandRecurringPayments = (payments: Payment[], students: Student[] = []): Payment[] => {
  const expanded: Payment[] = []
  const studentsById = new Map<string, Student>()
  
  for (const student of students) {
    studentsById.set(student.id, student)
  }
  
  for (const payment of payments) {
    if (payment.recurrence && payment.recurrence !== 'ONCE') {
      expanded.push(...generateRecurringPayments(payment, studentsById))
    } else {
      expanded.push(payment)
    }
  }

  return expanded
}

File: backend/app/services/payment_recurrence.py
"""Payment recurrence service - generates recurring payments"""
import uuid
from datetime import date, timedelta
from dateutil.relativedelta import relativedelta
from app.models import Payment
from app.schemas.payments import PaymentResponse


def generate_recurring_payments(
    base_payment: Payment,
    teacher_id: str,
    student_id: str,
    amount: float,
    recurrence: str,
    due_date: date,
    recurrence_end_date: date = None,
    **kwargs
) -> list[Payment]:
    """
    Gera pagamentos recorrentes a partir de um pagamento base
    
    Args:
        base_payment: Pagamento base com recorrência configurada
        teacher_id: ID do professor
        student_id: ID do aluno
        amount: Valor do pagamento
        recurrence: Tipo de recorrência (WEEKLY, BIWEEKLY, MONTHLY, ONCE)
        due_date: Data de vencimento inicial
        recurrence_end_date: Até quando repetir (ex: quando aluno sair da turma)
        **kwargs: Outros campos do pagamento
    
    Returns:
        Lista de objetos Payment gerados
    """
    if not recurrence or recurrence == 'ONCE':
        payment = Payment(
            id=str(uuid.uuid4()),
            teacher_id=teacher_id,
            student_id=student_id,
            amount=amount,
            due_date=due_date,
            recurrence=recurrence,
            recurrence_end_date=recurrence_end_date,
            **kwargs
        )
        return [payment]
    
    payments = []
    current_date = due_date
    
    # Se não há data de fim, assume 12 meses a partir de agora
    if not recurrence_end_date:
        recurrence_end_date = due_date + relativedelta(months=12)
    
    payment_index = 0
    
    while current_date <= recurrence_end_date:
        payment = Payment(
            id=str(uuid.uuid4()),
            teacher_id=teacher_id,
            student_id=student_id,
            amount=amount,
            due_date=current_date,
            recurrence=recurrence,
            recurrence_end_date=recurrence_end_date,
            status='PENDING' if current_date > date.today() else 'OVERDUE',
            **kwargs
        )
        payments.append(payment)
        
        # Calcula próxima data baseado na recorrência
        if recurrence == 'WEEKLY':
            current_date += timedelta(weeks=1)
        elif recurrence == 'BIWEEKLY':
            current_date += timedelta(weeks=2)
        elif recurrence == 'MONTHLY':
            current_date += relativedelta(months=1)
        
        payment_index += 1
        
        # Segurança: máximo 100 recorrências
        if payment_index > 100:
            break
    
    return payments


def expand_recurring_payments(payments: list[Payment]) -> list[Payment]:
    """
    Expande todos os payments com recorrência
    
    Args:
        payments: Lista de payments potencialmente com recorrência
    
    Returns:
        Lista expandida com todos os pagamentos gerados
    """
    expanded = []
    
    for payment in payments:
        if payment.recurrence and payment.recurrence != 'ONCE':
            expanded.extend(generate_recurring_payments(
                base_payment=payment,
                teacher_id=payment.teacher_id,
                student_id=payment.student_id,
                amount=payment.amount,
                recurrence=payment.recurrence,
                due_date=payment.due_date,
                recurrence_end_date=payment.recurrence_end_date,
                payment_date=payment.payment_date,
                payment_method=payment.payment_method,
                description=payment.description,
                status=payment.status,
                reference=payment.reference,
                notes=payment.notes,
            ))
        else:
            expanded.append(payment)
    
    return expanded
